### PR TITLE
Use newuidmap/newgidmap as fakeroot fallback

### PR DESCRIPTION
**Pull request by cclerget**
**Description of the Pull Request (PR):**

- Use newuidmap/newgidmap as fakeroot fallback for unprivileged installation.
- Fakeroot configuration is now handled by /etc/subxid files.
- Introduce new SINGULARITY_SUID_INSTALL constant to
ease setup detection and use starter binary accordingly.
- Fix fakeroot issues:
  - fix home directory (previously used user home directory)
  - HOME environment variable while joining an instance

**This fixes or addresses the following GitHub issues:**

- Related to discussion in #3891

**Before submitting a PR, make sure you have done the following:**

- Read the [Guidelines for Contributing](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md), and this PR conforms to the stated requirements.
- Added changes to the [CHANGELOG](https://github.com/sylabs/singularity/blob/master/CHANGELOG.md) if necessary according to the [Contribution Guidelines](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md)
- Added tests to validate this PR and tested this PR locally with a `make testall`
- Based this PR against the appropriate branch according to the [Contribution Guidelines](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md)
- Added myself as a contributor to the [Contributors File](https://github.com/sylabs/singularity/blob/master/CONTRIBUTORS.md)

Attn: @singularity-maintainers

### CODE REVIEW

File: internal/pkg/instance/instance_linux.go [hunk 1]
```diff
@@ -9,12 +9,12 @@ import (
 	"encoding/json"
 	"fmt"
 	"os"
-	"os/user"
 	"path/filepath"
 	"regexp"
 	"strings"
 	"syscall"
```
Comment (WestleyK) [inline internal/pkg/instance/instance_linux.go:16]: Copyright in this file should be `2018-2019`

File: internal/pkg/runtime/engines/singularity/prepare_linux.go [hunk 2]
```diff
@@ -21,7 +21,7 @@ import (
 
 	specs "github.com/opencontainers/runtime-spec/specs-go"
 	"github.com/sylabs/singularity/internal/pkg/buildcfg"
-	"github.com/sylabs/singularity/internal/pkg/fakeroot"
+	fakerootUtil "github.com/sylabs/singularity/internal/pkg/fakeroot"
```
Comment (WestleyK) [inline internal/pkg/runtime/engines/singularity/prepare_linux.go]: I thought I remember go[fmt,lint,vet] said package name should not have caps? Does that apply here?

File: internal/pkg/util/user/identity.go [hunk 3]
```diff
@@ -8,8 +8,13 @@
 
 package user
 
-import "strconv"
-import osuser "os/user"
+import (
+	"os"
+	osuser "os/user"
+	"strconv"
+
```
Comment (WestleyK) [inline internal/pkg/util/user/identity.go:12]: Copyright should also be `2018-2019` here

File: internal/pkg/util/user/identity_test.go [hunk 4]
```diff
@@ -9,6 +9,7 @@
 package user
```
Comment (WestleyK) [inline internal/pkg/util/user/identity_test.go:7]: Same here, `2018-2019`

File: pkg/util/namespaces/user_linux.go [hunk 5]
```diff
@@ -58,3 +58,53 @@ func IsInsideUserNamespace(pid int) (bool, bool) {
...
+func HostUID() (int, error) {
+	const (
+		uidMap = "/proc/self/uid_map"
+	)
+
+	currentUID := os.Getuid()
+
+	f, err := os.Open(uidMap)
+	if err != nil {
+		return 0, fmt.Errorf("failed to read %s: %s", uidMap, err)
```
Comment (WestleyK) [inline pkg/util/namespaces/user_linux.go]:
```suggestion
		return 0, fmt.Errorf("failed to read: %s: %s", uidMap, err)
```

File: pkg/util/namespaces/user_linux.go [hunk 6]
```diff
@@ -58,3 +58,53 @@ func IsInsideUserNamespace(pid int) (bool, bool) {
...
 	return insideUserNs, setgroupsAllowed
 }
+
+// HostUID attempts to find the original host UID if the current
+// process is running inside a user namespace, if it doesn't it
+// simply returns the current UID
+func HostUID() (int, error) {
+	const (
+		uidMap = "/proc/self/uid_map"
+	)
```
Comment (WestleyK) [inline pkg/util/namespaces/user_linux.go]: Should this section be simplified like?

```go
const uidMap = "/proc/self/uid_map"
```

File: pkg/util/namespaces/user_unsupported.go [hunk 7]
```diff
@@ -0,0 +1,24 @@
+// Copyright (c) 2018, Sylabs Inc. All rights reserved.
```
Comment (WestleyK) [inline pkg/util/namespaces/user_unsupported.go]:
```suggestion
// Copyright (c) 2019, Sylabs Inc. All rights reserved.
```

**Review by WestleyK — COMMENTED** [review]
Looks good, just a couple small suggestions :+1:

File: internal/pkg/runtime/engines/singularity/prepare_linux.go (hunk 2 above)
Comment (cclerget) [reply internal/pkg/runtime/engines/singularity/prepare_linux.go]: Ah yeah I forgot this rule, will fix that. Thanks !

**Review by WestleyK — APPROVED** [review]
Looks good to me! Thanks :+1:

**Comment by DrDaveD** [comment]
@cclerget I tried it out with one user, both setuid and not setuid, and didn't find any unexpected behavior.  Very cool.

**Pull request merged**